core/client/fs/src/test/java/alluxio/client/block/stream/SharedGrpcDataReaderTest.java

```diff
@@ -0,0 +1,421 @@
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.atLeastOnce;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.powermock.api.mockito.PowerMockito.when;
+
+import alluxio.ClientContext;
+import alluxio.ConfigurationTestUtils;
+import alluxio.client.file.FileSystemContext;
+import alluxio.collections.ConcurrentHashSet;
+import alluxio.grpc.Chunk;
+import alluxio.grpc.ReadRequest;
+import alluxio.grpc.ReadResponse;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.util.io.BufferUtils;
+import alluxio.wire.WorkerNetAddress;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Throwables;
+import com.google.protobuf.ByteString;
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.StreamObserver;
+import io.netty.buffer.ByteBuf;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({FileSystemContext.class, WorkerNetAddress.class})
+public final class SharedGrpcDataReaderTest {
+  private static final int CHUNK_SIZE = 1024;
+  private static final ExecutorService EXECUTOR = Executors.newFixedThreadPool(4);
+  private static final Random RANDOM = new Random();
+  private static final long BLOCK_ID = 1L;
+
+  private FileSystemContext mContext;
+  private WorkerNetAddress mAddress;
+  private BlockWorkerClient mClient;
+  private ClientCallStreamObserver<ReadRequest> mRequestObserver;
+
+  @Before
+  public void before() throws Exception {
+    mContext = PowerMockito.mock(FileSystemContext.class);
+    when(mContext.getClientContext())
+        .thenReturn(ClientContext.create(ConfigurationTestUtils.defaults()));
+    when(mContext.getClusterConf()).thenReturn(ConfigurationTestUtils.defaults());
+    mAddress = mock(WorkerNetAddress.class);
+    mClient = mock(BlockWorkerClient.class);
+    mRequestObserver = mock(ClientCallStreamObserver.class);
+    when(mContext.acquireBlockWorkerClient(mAddress)).thenReturn(
+        new NoopClosableResource<>(mClient));
+    when(mClient.readBlock(any(StreamObserver.class))).thenReturn(mRequestObserver);
+    when(mRequestObserver.isReady()).thenReturn(true);
```

Comment:
in general for unit test, I would not rely on mockito because the UT code becomes hard to maintain. For example, if we change the internal implementation of context or grpc streaming (which are totally internal and transparent to the class we are testing), it requires a lot of change here.

Here is how I would write this unit test. We can create a Mock class extending `BufferCachingGrpcDataReader`, and pass this mock reader to a `SharedGrpcDataReader` instance.  In this mock reader, we can return certain data, e.g., predefined for a block. And check the results on `SharedGrpcDataReader`.

This design pattern to inject a mock argument is more maintainable. Because you can focus on the logic of `SharedGrpcDataReader` that you want to test, without worrying about future change of `BufferCachingGrpcDataReader`. Let me know if I make myself clear. I am happy to write some example code too if that helps.